File: interview/Cracking the Coding Interview/chapter 2/2.5/c++/2.5.cpp
#include <iostream>

using namespace std;

template <typename T>
class Node {
	public:
	Node(){}
	Node(T data) { 
		this->data = data; 
		this->next = 0;
	}
	
	Node(T data, Node<T>* next){
		this->data = data;
		this->next = next;
	}
	
	T data;
	Node<T>* next;
};


template <typename T>
class LinkedList {
	public:
	LinkedList() { this->head = 0;}
	LinkedList(Node<T>* head) { this->head = head;}
	void insert(T data){
		Node<T> * n = new Node<T>(data,head);
		head = n;
	}
	
	Node<T>* search(T data){
		Node<T>* curr = head;
		
		while (curr && curr->data != data){
			curr = curr->next;
		}
		
		return curr;
	}
	
	void print(){
		Node<T>* curr = head;
		while (curr){
			cout << curr->data << " ";
			curr = curr->next;
		}
		
		cout << endl;
	}
	
	Node<T>* head;
	
};


LinkedList<int> add(LinkedList<int> list1, LinkedList<int> list2){
	LinkedList<int> sum = LinkedList<int>();
	
	int n1, n2;
	Node<int>* N1, * N2;
	
	N1 = list1.head;
	N2 = list2.head;
	
	int carry = 0;
	int s;
	while (N1 && N2){
		n1 = N1->data;
		n2 = N2->data;
		s = n1 + n2 + carry;
		carry = s / 10;
		s = s % 10;
		sum.insert(s);
		
		N1 = N1->next;
		N2 = N2->next;
	}
	
	while (N1){
		n1 = N1->data;
		s = n1 + carry;
		carry = s / 10;
		s = s % 10;
		sum.insert(s);
		
		N1 = N1->next;
	}
	
	while (N2) {
		n2 = N2->data;
		s = n2 + carry;
		carry = s / 10;
		s = s % 10;
		sum.insert(s);
		
		N2 = N2->next;
	}
	
	return sum;
}

LinkedList<int> add_reverse(LinkedList<int> list1, LinkedList<int> list2){
	
}

int main(){
	LinkedList<int> l1 = LinkedList<int>(new Node<int>(7, new Node<int>(1, new Node<int>(6)))),
					l2 = LinkedList<int>(new Node<int>(5, new Node<int>(9, new Node<int>(2))));
	l1.print();
	l2.print();
	add(l1,l2).print();
	
	return 0;
}

File: interview/Cracking the Coding Interview/chapter 2/2.5/python/2.5.py
class Node(object):
    def __init__(self, data=None, next_node=None):
        self.data = data
        self.next_node = next_node

    def get_data(self):
        return self.data

    def get_next(self):
        return self.next_node

    def set_next(self, new_next):
        self.next_node = new_next

class LinkedList(object):
    def __init__(self, head=None):
        self.head = head

    def insert(self, data):
        new_node = Node(data)
        new_node.set_next(self.head)
        self.head = new_node

    def size(self):
        current = self.head
        count = 0
        while current:
            count += 1
            current = current.get_next()
        return count

    def __len__(self):
        return self.size()

    def search(self, data):
        current = self.head
        found = False
        while current and found is False:
            if current.get_data() == data:
                found = True
            else:
                current = current.get_next()
        if current is None:
            raise ValueError("Data not in list")
        return current

    def __contains__(self, data):
        return self.search(data)
        
    def delete(self, data):
        current = self.head
        previous = None
        found = False
        while current and found is False:
            if current.get_data() == data:
                found = True
            else:
                previous = current
                current = current.get_next()

        if current is None:
            raise ValueError("Data not in list")
        if previuos is None:
            self.head = current.get_next()
        else:
            previous.set_next(current.get_next())

    def __repr__(self):
        s = ''
        current = self.head
        while current:
            s += str(current.get_data()) + ','
            current = current.get_next()
        return s[:-1]

def add(n1, n2):
    S = LinkedList()
    curr_S = None
    curr_n1 = n1.head
    curr_n2 = n2.head

    carry = 0
    while curr_n1 != None and curr_n2 != None:
        s = curr_n1.get_data() + curr_n2.get_data() + carry
        
        if curr_S == None:
            S.insert(s % 10)
            curr_S = S.head
        else:
            N = Node(s % 10)
            curr_S.set_next(N)
            curr_S = N
            
        carry = int(s/10)
        
        curr_n1 = curr_n1.get_next()
        curr_n2 = curr_n2.get_next()

    if curr_n1 == None:
        while curr_n2 != None:
            s = curr_n2.get_data() + carry
            N = Node(s % 10)
            curr_S.set_next(N)
            curr_S = N
            carry = int(s/10)
            
            curr_n2 = curr_n2.get_next()
            
    elif curr_n2 == None:
            s = curr_n1.get_data() + carry
            N = Node(s % 10)
            curr_S.set_next(N)
            curr_S = N
            carry = int(s/10)
        
            curr_n1 = curr_n1.get_next()
            
    if carry > 0:
        curr_S.set_next(Node(carry))
        
    return S
##
##def add_forward(n1, n2):
##    S = LinkedList()
##    curr_S = None
##    curr_n1 = n1.head
##    curr_n2 = n2.head
##
##    carry = 0
##    while curr_n1 != None and curr_n2 != None:
##        s = curr_n1.get_data() + curr_n2.get_data() + carry
##        
##        S.insert(s % 10)            
##        carry = int(s/10)
##        
##        curr_n1 = curr_n1.get_next()
##        curr_n2 = curr_n2.get_next()
##
##    if curr_n1 == None:
##        while curr_n2 != None:
##            s = curr_n2.get_data() + carry
##            N = Node(s % 10)
##            curr_S.set_next(N)
##            curr_S = N
##            carry = int(s/10)
##            
##            curr_n2 = curr_n2.get_next()
##            
##    elif curr_n2 == None:
##            s = curr_n1.get_data() + carry
##            N = Node(s % 10)
##            curr_S.set_next(N)
##            curr_S = N
##            carry = int(s/10)
##        
##            curr_n1 = curr_n1.get_next()
##            
##    if carry > 0:
##        curr_S.set_next(Node(carry))
##        
##    return S

# 123 + 123 = 246
L1 = LinkedList(Node(3,Node(2,Node(1))))
L2 = LinkedList(Node(3,Node(2,Node(1))))
print(add(L1,L2))

# 617 + 295 = 912
L1 = LinkedList(Node(7,Node(1,Node(6))))
L2 = LinkedList(Node(5,Node(9,Node(2))))
print(add(L1,L2))

# 717 + 295 = 1012
L1 = LinkedList(Node(7,Node(1,Node(7))))
L2 = LinkedList(Node(5,Node(9,Node(2))))
print(add(L1,L2))

# 123 + 1081 = 1204
L1 = LinkedList(Node(3,Node(2,Node(1))))
L2 = LinkedList(Node(1,Node(8,Node(0,Node(1)))))
print(add(L1,L2))
